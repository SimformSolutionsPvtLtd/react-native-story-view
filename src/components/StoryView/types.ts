import type { RefObject } from 'react';
import type {
  ActivityIndicatorProps,
  ImageProps,
  ImageSourcePropType,
  ImageStyle,
  TextProps,
  TextStyle,
  View,
  ViewProps,
  ViewStyle,
} from 'react-native';
import type { GestureHandlerEvent } from 'react-native-reanimated/lib/typescript/reanimated2/hook';
import type {
  OnLoadData,
  OnProgressData,
  VideoProperties,
} from 'react-native-video';

export enum StroyTypes {
  Image = 'image',
  Video = 'video',
  Text = 'text',
}

export enum ClickPosition {
  Right = 'right',
  Left = 'left',
}

export enum ProgressState {
  Default = 0,
  InProgress = 1,
  Completed = 2,
  Paused = 3,
}

export enum StoryMode {
  StoryView = 'StoryView',
  SingleStory = 'SingleStory',
}

export interface CommonProps {
  images?: Array<string>;
  duration?: number | undefined;
  containerStyle?: ViewStyle;
  enableProgress?: boolean | undefined;
  imageStyle?: ImageStyle;
  barStyle?: BarStyleProps;
  maxVideoDuration?: number | undefined;
}

export interface BarStyleProps {
  barActiveColor?: string;
  barInActiveColor?: string;
  barHeight?: number;
}

export interface ProgressBarCommonProps {
  next?: () => void;
  pause: boolean;
  isLoaded: boolean;
  duration: number;
  currentIndex: number;
  barStyle?: BarStyleProps;
  progressBarStyle?: ViewStyle | undefined;
}

export interface ProgressBarProps extends ProgressBarCommonProps {
  active: number;
  index: number;
  length: number;
  storyType?: string;
  videoDuration: number[];
  setVideoDuration: (duration: number[]) => void;
  storyIndex: number;
  currentUserIndex?: number;
}

export interface ProgressBarsProps extends ProgressBarCommonProps {
  stories: StoryType[];
  currentStory: Object;
  videoDuration: number[];
  setVideoDuration: (duration: number[]) => void;
  length: Array<number>;
  progress: Object;
  storyIndex: number;
  index?: number;
}

export interface StoryViewProps {
  pause?: boolean;
  onVideoLoaded?: (arg0: OnLoadData) => void;
  videoDuration: number[];
  onVideoProgress?: (progressData?: OnProgressData) => void;
  onVideoEnd?: () => void;
  onImageLoaded?: () => void;
  stories: StoryType[];
  showSourceIndicator?: boolean;
  sourceIndicatorProps?: ActivityIndicatorProps;
  videoProps?: VideoProperties;
  index?: number;
  storyIndex?: number;
}

export interface StoryViewProps extends CommonProps {
  progressIndex: number;
  viewRef?: RefObject<View>;
}

export interface ProgressiveImageProps {
  thumbnailSource: {
    uri: string;
  };
  imgSource: {
    uri: string;
  };
  viewStyle?: ImageStyle;
  props?: any;
  onImageLoaded?: () => void;
}

export interface CallbackProps {
  userStories?: StoriesType | undefined;
  story?: StoryType[] | undefined;
  progressIndex: number;
  userStoryIndex?: number | undefined;
}

export interface StoryContainerProps extends CommonProps {
  stories: StoryType[];
  userStories?: StoriesType;
  visible?: boolean | undefined;
  index?: number | undefined;
  userStoryIndex?: number | undefined;
  storyIndex?: number | undefined;
  isShowReply?: boolean | undefined;
  renderHeaderComponent?: (callback: CallbackProps) => JSX.Element;
  renderFooterComponent?: (callback: CallbackProps) => JSX.Element;
  renderIndicatorComponent?: () => JSX.Element;
  userProfile?: UserProps | undefined;
  footerView?: FooterViewProps | undefined;
  onComplete?: (viewedStories?: Array<boolean[]>) => void;
  renderCustomView?: (callback: CallbackProps) => JSX.Element;
  backgroundColor?: string;
  style?: ViewStyle;
  progressIndex?: number | undefined;
  headerViewProps?: ViewProps;
  customViewProps?: ViewProps;
  footerViewProps?: ViewProps;
  progressViewProps?: ViewProps;
  storyContainerViewProps?: ViewProps;
  showSourceIndicator?: boolean;
  sourceIndicatorProps?: ActivityIndicatorProps;
  videoProps?: VideoProperties;
  onChangePosition?: (
    progressIndex: number,
    userStoryIndex?: number
  ) => void | undefined;
  previousStory?: () => void | undefined;
  nextStory?: () => void;
  customViewStyle?: ViewStyle;
  headerStyle?: ViewStyle;
  footerStyle?: ViewStyle;
  gestureHandler?: (e: GestureHandlerEvent<any>) => void;
}

export interface UserProps {
  containerStyle?: ViewStyle | undefined;
  userImage?: ImageSourcePropType | undefined;
  userName?: string | undefined;
  userMessage?: string | undefined;
  imageArrow?: ImageSourcePropType | undefined;
  onImageClick?: Function | undefined;
  onClosePress?: Function | undefined;
  userImageProps?: ImageProps | undefined;
  closeIconProps?: ImageProps | undefined;
  userNameProps?: TextProps | undefined;
  userMessageProps?: TextProps | undefined;
  userImageStyle?: ImageStyle;
  rootStyle?: ViewStyle;
  userNameStyle?: TextStyle;
  userMessageStyle?: TextStyle;
  closeIconStyle?: ImageStyle;
  customCloseButton?: any;
}

export interface FooterViewProps {
  isShowReply: boolean | undefined;
  onReplyTextChange: Function | undefined;
  onReplyButtonClick: Function | undefined;
}

export type StoryType = {
  id: number;
  url?: string;
  thumbnail?: string;
  type?: string | 'image' | 'video' | 'text';
  duration?: number;
  isReadMore?: boolean;
  isSeen?: boolean;
  isPaused?: boolean;
  created?: string;
  storyId?: number;
  title?: string;
};

export type StoriesType = {
  username?: string;
  profile?: string;
  title?: string;
  id: number;
  stories: StoryType[];
};

export type StoryRef = {
  pause: (pause: boolean) => void;
  handleLongPress: (visibility: boolean) => void;
  viewedStories: boolean[];
};
