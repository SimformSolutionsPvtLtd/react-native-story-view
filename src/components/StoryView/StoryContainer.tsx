import React, {
  forwardRef,
  useEffect,
  useImperativeHandle,
  useRef,
  useState,
} from 'react';
import {
  KeyboardAvoidingView,
  SafeAreaView,
  TouchableOpacity,
  View,
} from 'react-native';
import {
  GestureHandlerRootView,
  PanGestureHandler,
} from 'react-native-gesture-handler';
import Animated from 'react-native-reanimated';
import { Metrics } from '../../theme';
import ProgressView from './ProgressView';
import StoryView from './StoryView';
import { useStoryContainer } from './hooks';
import styles from './styles';
import {
  ClickPosition,
  StoryContainerProps,
  StoryMode,
  StoryRef,
} from './types';

const StoryContainer = forwardRef<StoryRef, StoryContainerProps>(
  (
    {
      renderHeaderComponent,
      renderFooterComponent,
      renderCustomView,
      userStories,
      enableProgress = true,
      headerViewProps,
      customViewProps,
      footerViewProps,
      progressViewProps,
      storyContainerViewProps,
      gestureHandler,
      ...props
    }: StoryContainerProps,
    ref
  ) => {
    const viewedStories = useRef<boolean[]>(
      Array(props?.stories?.length)
        .fill(props?.stories)
        .map((item, index) => item?.[index]?.isSeen ?? false)
    );

    const {
      progressIndex,
      isPause,
      setPause,
      isLoaded,
      duration,
      opacity,
      onImageLoaded,
      onVideoLoaded,
      changeStory,
      setLoaded,
      setDuration,
      videoDuration,
      onVideoProgress,
      onVideoEnd,
      onArrowClick,
      onStoryPressHold,
      isKeyboardVisible,
      setVideoDuration,
      onStoryPressRelease,
      setVisibleElements,
      rootStyle,
      containerStyle,
    } = useStoryContainer(props, viewedStories);

    const viewRef = useRef<View>(null);

    useImperativeHandle(ref, () => ({
      pause: (pause: boolean) => {
        if (props?.index === props?.userStoryIndex) {
          setPause(pause);
        }
      },
      handleLongPress: (visibility: boolean) => {
        if (props?.index === props?.userStoryIndex) {
          setVisibleElements(!visibility);
          setPause(visibility);
        }
      },

      viewedStories: viewedStories.current,
    }));

    useEffect(() => {
      setLoaded(false);
      setDuration(
        props.stories?.[progressIndex]?.duration ?? Metrics.defaultDuration
      );
    }, [progressIndex, props.stories, setDuration, setLoaded]);

    const storyMode: StoryMode =
      props?.userStoryIndex !== undefined
        ? StoryMode.StoryView
        : StoryMode.SingleStory;
    const [viewHeight, setViewHeight] = useState<number>(0);
    const bottom =
      storyMode === StoryMode.SingleStory
        ? isKeyboardVisible
          ? viewHeight
          : 0
        : 0;

    const storyViewContent = () => {
      return (
        <GestureHandlerRootView style={styles.rootViewStyle}>
          <PanGestureHandler
            activateAfterLongPress={200}
            onGestureEvent={gestureHandler}>
            <Animated.View
              style={props?.containerStyle ?? styles.parentView}
              {...storyContainerViewProps}>
              <View
                onLayout={({ nativeEvent }) => {
                  if (isKeyboardVisible) return;
                  const { height } = nativeEvent.layout;
                  viewRef?.current?.setNativeProps({ height });
                }}
                style={props?.containerStyle ?? styles.parentView}
                {...storyContainerViewProps}>
                <TouchableOpacity
                  activeOpacity={1}
                  delayLongPress={200}
                  onPress={(e: { nativeEvent: any }) =>
                    changeStory(e.nativeEvent)
                  }
                  onLongPress={onStoryPressHold}
                  onPressOut={onStoryPressRelease}>
                  <StoryView
                    viewRef={viewRef}
                    duration={duration}
                    onVideoLoaded={onVideoLoaded}
                    onImageLoaded={onImageLoaded}
                    progressIndex={progressIndex}
                    videoDuration={videoDuration}
                    onVideoEnd={onVideoEnd}
                    onVideoProgress={onVideoProgress}
                    pause={isPause}
                    index={props?.index ?? 0}
                    storyIndex={props?.userStoryIndex ?? 0}
                    stories={props.stories}
                    imageStyle={props.imageStyle}
                    videoProps={props?.videoProps}
                    sourceIndicatorProps={props?.sourceIndicatorProps}
                    showSourceIndicator={props?.showSourceIndicator ?? true}
                  />
                </TouchableOpacity>
                {enableProgress && (
                  <View
                    style={[styles.progressView, { opacity }]}
                    {...progressViewProps}>
                    <ProgressView
                      next={() => onArrowClick(ClickPosition.Right)}
                      isLoaded={isLoaded}
                      duration={duration}
                      storyIndex={props?.userStoryIndex ?? 0}
                      currentIndex={progressIndex}
                      setVideoDuration={setVideoDuration}
                      index={props?.index ?? 0}
                      videoDuration={videoDuration ?? 0}
                      pause={enableProgress && isPause}
                      stories={props?.stories}
                      barStyle={props?.barStyle}
                      currentStory={props?.stories[progressIndex]}
                      length={props?.stories?.map((_, i) => i)}
                      progress={{ id: progressIndex }}
                    />
                  </View>
                )}
                {renderHeaderComponent && (
                  <View
                    style={[
                      styles.topView,
                      props?.headerStyle ?? {},
                      { opacity },
                    ]}
                    {...headerViewProps}>
                    <>
                      {renderHeaderComponent?.({
                        userStories,
                        story: props?.stories,
                        progressIndex,
                        userStoryIndex: props?.userStoryIndex,
                      })}
                    </>
                  </View>
                )}
                {renderCustomView && (
                  <View
                    style={[
                      styles.customView,
                      props?.customViewStyle ?? {},
                      { opacity },
                    ]}
                    {...customViewProps}>
                    <>
                      {renderCustomView?.({
                        userStories,
                        story: props?.stories,
                        progressIndex,
                        userStoryIndex: props?.userStoryIndex,
                      })}
                    </>
                  </View>
                )}
              </View>
            </Animated.View>
          </PanGestureHandler>
          {renderFooterComponent && (
            <Animated.View
              style={[
                styles.bottomView,
                props?.footerStyle ?? {},
                { opacity, bottom },
              ]}
              onLayout={({ nativeEvent }) => {
                setViewHeight(nativeEvent.layout.height);
              }}
              {...footerViewProps}>
              <>
                {renderFooterComponent?.({
                  userStories,
                  story: props?.stories,
                  progressIndex,
                  userStoryIndex: props?.userStoryIndex,
                })}
              </>
            </Animated.View>
          )}
        </GestureHandlerRootView>
      );
    };

    return (
      <SafeAreaView style={rootStyle}>
        <KeyboardAvoidingView
          style={containerStyle}
          keyboardVerticalOffset={Metrics.keyboardVerticalOffset}
          behavior={'padding'}>
          {props.visible && storyViewContent()}
        </KeyboardAvoidingView>
      </SafeAreaView>
    );
  }
);

export default StoryContainer;
