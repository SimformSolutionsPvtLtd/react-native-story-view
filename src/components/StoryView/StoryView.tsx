import React, { useEffect, useRef, useState } from 'react';
import { ActivityIndicator, useWindowDimensions, View } from 'react-native';
import Video, { OnBufferData, OnLoadData } from 'react-native-video';
import { Colors, Metrics } from '../../theme';
import ProgressiveImage from './ProgressiveImage';
import styles from './styles';
import { StoryViewProps, StroyTypes } from './types';

const BUFFER_TIME = 1000 * 60;

const StoryView = (props: StoryViewProps) => {
  const [loading, setLoading] = useState(true);
  const [buffering, setBuffering] = useState(true);
  const source = props?.stories?.[props?.progressIndex];
  const videoRef = useRef<Video>(null);
  const videoData = useRef<OnLoadData>();
  const isCurrentIndex = props?.index === props?.storyIndex;

  useEffect(() => {
    if (props?.index === props?.storyIndex) {
      videoRef?.current?.seek(0);
    }
  }, [props?.storyIndex, props?.index]);

  const onLoadStart = () => {
    setLoading(true);
  };

  const loadVideo = () => {
    if (isCurrentIndex) {
      if (videoData.current === undefined) return;
      setLoading(false);
      setBuffering(false);
      props?.onVideoLoaded?.(videoData.current);
    }
  };

  const onBuffer = (data: OnBufferData) => {
    setBuffering(data.isBuffering);
  };

  const { height, width } = useWindowDimensions();

  return (
    <View style={[styles.divStory, { height, width }]} ref={props?.viewRef}>
      {source?.type === StroyTypes.Image ? (
        <ProgressiveImage
          viewStyle={props?.imageStyle ?? styles.imgStyle}
          imgSource={{ uri: source.url ?? '' }}
          thumbnailSource={{ uri: source.url ?? '' }}
          onImageLoaded={props.onImageLoaded}
        />
      ) : (
        isCurrentIndex && (
          <>
            <Video
              ref={videoRef}
              resizeMode="contain"
              paused={props.pause || loading}
              source={{
                uri: source?.url,
              }}
              onEnd={props?.onVideoEnd}
              onError={(_error: any) => {
                setLoading(false);
              }}
              onProgress={data => {
                if (isCurrentIndex) {
                  props?.onVideoProgress?.(data);
                }
              }}
              bufferConfig={{
                minBufferMs: BUFFER_TIME,
                bufferForPlaybackMs: BUFFER_TIME,
                bufferForPlaybackAfterRebufferMs: BUFFER_TIME,
              }}
              onBuffer={onBuffer}
              onLoadStart={onLoadStart}
              onLoad={(item: OnLoadData) => {
                videoData.current = item;
                !Metrics.isIOS && loadVideo();
              }}
              onReadyForDisplay={loadVideo}
              style={styles.contentVideoView}
              {...props?.videoProps}
            />
            {(loading || buffering) && props?.showSourceIndicator && (
              <ActivityIndicator
                animating
                pointerEvents="none"
                color={Colors.loaderColor}
                size="small"
                style={styles.loaderView}
                {...props?.sourceIndicatorProps}
              />
            )}
          </>
        )
      )}
    </View>
  );
};

export default StoryView;
