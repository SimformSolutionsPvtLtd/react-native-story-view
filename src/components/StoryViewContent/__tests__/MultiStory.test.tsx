import React from 'react';
import { render } from '@testing-library/react-native';
import StoryView from '../StoryView';

jest.useFakeTimers();

describe('StoryView Component', () => {
  it('Match Snapshot', () => {
    const { toJSON } = render(<StoryView stories={[]} />);
    expect(toJSON()).toMatchSnapshot();
  });
});
