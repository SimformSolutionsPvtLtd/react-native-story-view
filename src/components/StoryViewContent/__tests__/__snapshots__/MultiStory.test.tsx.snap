// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoryView Component Match Snapshot 1`] = `
<View>
  <RCTScrollView
    data={Array []}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={true}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    showsHorizontalScrollIndicator={false}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View />
  </RCTScrollView>
</View>
`;
